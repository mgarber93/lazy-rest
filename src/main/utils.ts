import {EndpointCallPlan, THttp} from '../models/endpoint'


export function parseCalls(plan: string): EndpointCallPlan[] {
  const endpoints = plan
    .split('\n')
    .filter(str => str.startsWith('API calling'))
    .map(str => str.split(":", 2)[1].trim())
  
  const httpCalls: EndpointCallPlan[] = endpoints.map(endpoint => {
    const [method, path, ...rest] = endpoint.split(' ')
    return {method: method.toUpperCase() as THttp, path, background: rest.join(' ')}
  })
  
  return httpCalls
}
