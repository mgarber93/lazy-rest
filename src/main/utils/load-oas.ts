import {OpenApiSpec} from '../../models/open-api-spec'
import windowSender from './window-sender'

export async function loadOas(): Promise<OpenApiSpec[]> {
  try {
    const response = await windowSender.asyncSend("load-oas")
    return response
  } catch (error) {
    console.error(error)
  }
}