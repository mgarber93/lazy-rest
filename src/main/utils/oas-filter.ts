import {Endpoint, OpenApiSpec} from '../../models/open-api-spec'
import {HttpRequestPlan} from '../../models/conversation'

function setEndpointDescription(object: Record<string, any>, path: string, key: string, value: any) {
  if (!(path in object)) {
    object[path] = {}
  }
  object[path][key] = value.replace('\n', ' ').trim()
}

/**
 * gets the description of paths in the specification
 * It does not read the description of the oas overall, which is also available
 * @param oasSpec
 */
export function oasToDescriptions(oasSpec: OpenApiSpec): object {
  const spec = {}
  for (const key in oasSpec.paths) {
    const endpoint = oasSpec.paths[key]
    if (endpoint.get?.description) {
      setEndpointDescription(spec, key, 'get', endpoint.get.description)
    }
    if (endpoint.put?.description) {
      setEndpointDescription(spec, key, 'put', endpoint.put.description)
    }
    if (endpoint.patch?.description) {
      setEndpointDescription(spec, key, 'patch', endpoint.patch.description)
    }
    if (endpoint.post?.description) {
      setEndpointDescription(spec, key, 'post', endpoint.post.description)
    }
    if (endpoint.delete?.description) {
      setEndpointDescription(spec, key, 'delete', endpoint.delete.description)
    }
  }
  return spec
}

export function fuzzyMatch(a: string, b: string): boolean {
  const aSegments = a.split("/")
  const bSegments = b.split("/")
  
  if (aSegments.length !== bSegments.length) {
    return false
  }
  
  for (let i = 0; i < aSegments.length; i++) {
    const aSegment = aSegments[i]
    const bSegment = bSegments[i]
    // if a segment is surrounded by curly braces, acts like a wild card
    if (aSegment.startsWith("{") && aSegment.endsWith("}")) {
      continue
    }
    if (bSegment.startsWith("{") && bSegment.endsWith("}")) {
      continue
    }
    if (aSegments[i] !== bSegments[i]) {
      return false
    }
  }
  return true
}

export function treeShake(oasSpec: OpenApiSpec, plans: HttpRequestPlan[]) {
  const spec = {} as Record<string, any>
  for (const endpointPath in oasSpec.paths) {
    for (const plan of plans) {
      if (fuzzyMatch(endpointPath, plan.path)) {
        const matched = oasSpec.paths[endpointPath]
        const verb = plan.method.toLowerCase()
        spec[endpointPath] = {}
        spec[endpointPath][verb] = matched[verb as keyof Endpoint]
      }
    }
  }
  return spec
}