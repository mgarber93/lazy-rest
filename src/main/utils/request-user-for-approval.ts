import windowSender from './window-sender'
import {Approvable, ApprovalResponse, approvalResponseIsApproved} from '../../models/approvable'


export async function requestUserForApproval(request: Approvable, ...args: any[]): Promise<ApprovalResponse> {
  const response = await windowSender.asyncSend('approval', request, ...args)
  if (!approvalResponseIsApproved(response)) {
    throw new Error('unapproved')
  }
  return response
}