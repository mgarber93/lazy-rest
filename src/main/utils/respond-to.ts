import windowSender from './window-sender'
import {AuthoredContent} from '../../models/content'
import {HttpRequestPlan} from '../../models/http-request-plan'


export async function respondTo(chatId: string, author: string): Promise<AuthoredContent> {
  const content = await windowSender.asyncSend('respond-to', chatId)
  return content
}

export async function presentCallingPlan(chatId: string, calls: HttpRequestPlan[]) {
  windowSender.send('calling-plan', chatId, calls)
}