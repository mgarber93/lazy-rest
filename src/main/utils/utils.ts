import {OpenAPI} from 'openapi-types'
import {HttpRequestPlan, THttp} from '../../models/conversation'


export function parseCalls(plan: string, apiSpec: OpenAPI.Document): HttpRequestPlan[] {
  const endpoints = plan
    .split('\n')
    .filter(str => str.startsWith('API calling'))
    .map(str => str.split(":", 2)[1].trim())
  
  const httpCalls = endpoints.map(endpoint => {
    const [method, path, ...rest] = endpoint.split(' ')
    const mappedMethod = method.toUpperCase() as THttp
    return {
      baseUrl:
      path,
      method: mappedMethod,
      body: {},
      headers: {},
    } as HttpRequestPlan
  })
  
  return httpCalls
}
