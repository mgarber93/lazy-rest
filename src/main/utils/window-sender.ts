import {v4} from 'uuid'

export type TChannel = "message-delta" | "load-oas" | 'callback' | 'approval'

export type TSender = (eventName: string, ...args: any[]) => void

export class WindowSender {
  private _sender: TSender | null = null
  private promiseMap = new Map<string, (value: unknown) => void>()
  constructor(private _queue: { eventName: TChannel, args: any[] }[] = []) {
  }
  hasFinishedLoading(sender: TSender) {
    this._sender = sender
    for (const message of this._queue) {
      this._sender(message.eventName, ...message.args)
    }
  }
  private sendOrQueue(eventName: TChannel, args: any[]) {
    if (this._sender) {
      this._sender(eventName, ...args)
    } else {
      this._queue.push({eventName, args: [...args]})
    }
  }
  send(eventName: TChannel, ...args: any[]): void {
    return this.sendOrQueue(eventName, args)
  }

  /**
   * Async send allows the renderer process (received by sender fxn) to complete some async process and then return
   * value of type T using the callback method defined on this class.
   * @param eventName
   * @param args
   */
  asyncSend<T = any>(eventName: TChannel, ...args: any[]): Promise<T> {
    const id = v4()
    return new Promise((resolve, reject) => {
      if (this.promiseMap.has(id)) {
        throw new Error(`${id} is already registered (do we need to set a random seed?)`)
      }
      this.promiseMap.set(id, resolve)

      const nextArgs = [id, ...args]
      if (this._sender) {
         this._sender(eventName, ...nextArgs)
      } else {
        this._queue.push({eventName, args: nextArgs})
      }
    })
  }
  callback(id:string, response: any) {
    const resolve = this.promiseMap.get(id)
    if (resolve) {
      this.promiseMap.delete(id)
      return resolve(response)
    } else {
      throw new Error(`No promise found for id: ${id}`)
    }
  }
}

export default new WindowSender()