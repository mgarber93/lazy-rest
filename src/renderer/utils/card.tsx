export const cardEffect = "rounded-xl px-4 py-2 bg-zinc-50 dark:bg-zinc-800 dark:shadow dark:shadow-zinc-900/2 drop-shadow mb-2 shadow-zinc-300 dark:shadow-zinc-950 border-2 border-transparent hover:border-zinc-100 hover:border-opacity-100 dark:hover:border-transparent box-border transition-all duration-1000"
