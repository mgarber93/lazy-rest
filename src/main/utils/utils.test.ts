import {parseCalls} from './utils'
import {OpenAPI} from 'openapi-types'

test('parseCalls', () => {
  const test = 'Background: No background  \n' +
    'User query: What are the most popular songs by the artist altrice?  \n' +
    'API calling 1: GET /search to search for the artist "altrice"  \n' +
    'API response: altrice\'s artist_id is xyz123abc  \n' +
    'API calling 2: GET /artists/xyz123abc/top-tracks to get the most popular songs by altrice  \n' +
    'API response: The most popular songs by altrice are "Song A", "Song B", "Song C"  \n' +
    'Instruction: No further API calls needed.'
  const actual = parseCalls(test, {} as OpenAPI.Document)
  expect(actual).toMatchSnapshot()
})
