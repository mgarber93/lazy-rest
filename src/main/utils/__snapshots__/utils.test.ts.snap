// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseCalls 1`] = `
[
  {
    "baseUrl": "/search",
    "body": {},
    "headers": {},
    "method": "GET",
  },
  {
    "baseUrl": "/artists/xyz123abc/top-tracks",
    "body": {},
    "headers": {},
    "method": "GET",
  },
]
`;
